futures-util/src/future/poll_immediate.rs

```diff
@@ -0,0 +1,201 @@
+use super::assert_future;
+use crate::FutureExt;
+use core::pin::Pin;
+use futures_core::task::{Context, Poll};
+use futures_core::{FusedFuture, Future, Stream};
+
+/// Future for the [`poll_immediate`](poll_immediate()) function.
+#[derive(Debug, Clone)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct PollImmediate<T>(Option<T>);
+
+impl<T, F> Future for PollImmediate<F>
+where
+    F: Future<Output = T>,
+{
+    type Output = Option<T>;
+
+    #[inline]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<T>> {
+        // # Safety
+        // This is the only time that this future will ever be polled.
+        let inner =
+            unsafe { self.get_unchecked_mut().0.take().expect("PollOnce polled after completion") };
+        crate::pin_mut!(inner);
+        match inner.poll(cx) {
+            Poll::Ready(t) => Poll::Ready(Some(t)),
+            Poll::Pending => Poll::Ready(None),
+        }
+    }
+}
+
+impl<T: Future> FusedFuture for PollImmediate<T> {
+    fn is_terminated(&self) -> bool {
+        self.0.is_none()
+    }
+}
+
+/// Creates a stream that can be polled repeatedly until the future is done
+/// ```
+/// # futures::executor::block_on(async {
+/// use futures::task::Poll;
+/// use futures::{StreamExt, future, pin_mut};
+/// use future::FusedFuture;
+///
+/// let f = async { 1_u32 };
+/// pin_mut!(f);
+/// let mut r = future::poll_immediate(f);
+/// assert_eq!(r.next().await, Some(Poll::Ready(1)));
+///
+/// let f = async {futures::pending!(); 42_u8};
+/// pin_mut!(f);
+/// let mut p = future::poll_immediate(f);
+/// assert_eq!(p.next().await, Some(Poll::Pending));
+/// assert!(!p.is_terminated());
+/// assert_eq!(p.next().await, Some(Poll::Ready(42)));
+/// assert!(p.is_terminated());
+/// assert_eq!(p.next().await, None);
+/// # });
+/// ```
+impl<T, F> Stream for PollImmediate<F>
+where
+    F: Future<Output = T>,
+{
+    type Item = Poll<T>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        unsafe {
+            // # Safety
+            // We never move the inner value until it is done. We only get a reference to it.
+            let inner = &mut self.get_unchecked_mut().0;
+            let fut = match inner.as_mut() {
+                // inner is gone, so we can signal that the stream is closed.
+                None => return Poll::Ready(None),
+                Some(inner) => inner,
+            };
+            let fut = Pin::new_unchecked(fut);
+            Poll::Ready(Some(fut.poll(cx).map(|t| {
+                // # Safety
+                // The inner option value is done, so we need to drop it. We do it without moving it
+                // by using drop in place. We then write over the value without trying to drop it first
+                // This should uphold all the safety requirements of `Pin`
+                std::ptr::drop_in_place(inner);
+                std::ptr::write(inner, None);
+                t
```

Comment:
this is unsound because `inner` may panic

Reply:
I found the `set()` function in the abortable code.